### PR TITLE
Feat/dedup message

### LINKED ISSUE #76: De-duplicate messages based on message ID from Notify Server
Notify server now generates IDs for messages. These can be used to de-duplicate messages when the same message is sent to the relay. That's because even though they'd have different `payload.id`s  but the same `notifyMessageId`

### PR DESCRIPTION
**Pull request by devceline**
Resolves #76 

Dedup based on message ID coming from Notify server

Tested via new unit test

### CODE REVIEW

File: packages/notify-client/src/controllers/engine.ts [hunk 1]
```diff
@@ -622,6 +622,17 @@ export class NotifyEngine extends INotifyEngine {
 
       const currentMessages = this.client.messages.get(topic).messages;
 
+      const messageIdAlreadyReceived = Object.values(currentMessages).some(
```
Comment (bkrem) [inline packages/notify-client/src/controllers/engine.ts:625]: Looking at types (https://github.com/WalletConnect/notify-client-js/blob/main/packages/notify-client/src/types/client.ts#L248) isn't this record indexed by messageId already?

Given that `messages` is the most likely to be the biggest map the NotifyClient handles, I think it would be prudent to try and avoid the full double iteration from `Object.values(...).some(...)` here and instead check the object index for the messageId for the same outcome, which would be `O(1)` instead of `O(2n)`?

Something along the lines of:
```
const messageIdAlreadyReceived = currentMessages[messageClaims.msg.id] !== undefined
```

File: packages/notify-client/src/controllers/engine.ts [hunk 2]
```diff
@@ -622,6 +622,17 @@ export class NotifyEngine extends INotifyEngine {
...
+        (msg) => msg.message.id === messageClaims.msg.id
+      );
+
+      if (messageIdAlreadyReceived) {
+        this.client.logger.warn(
+          `[Notify] Message with id ${messageClaims.msg.id} already received`
```
Comment (bkrem) [inline packages/notify-client/src/controllers/engine.ts:631]: Just capturing in the message what action the client has taken in response to the edge case:

```suggestion
          `[Notify] Message with id ${messageClaims.msg.id} already received. Ignoring.`
```

File: packages/notify-client/test/clients.spec.ts [hunk 3]
```diff
@@ -841,5 +848,63 @@ describe("Notify", () => {
...
+
+        let messagesReceived = 0;
+
+        wallet.on("notify_message", () => {
+          messagesReceived++;
+        });
+
+        const now = Date.now();
+
+        await waitForEvent(() => Date.now() - now > 1_000);
```
Comment (bkrem) [inline packages/notify-client/test/clients.spec.ts:872]: Why the arbitrary duration? Will almost definitely start flaking eventually in my experience if it's not tied to some concrete event/state

File: packages/notify-client/test/clients.spec.ts [hunk 4]
```diff
@@ -841,5 +848,63 @@ describe("Notify", () => {
...
+          message = m;
+        });
+
+        await sendNotifyMessage(account, "Test");
+
+        await waitForEvent(() => receivedRealMessage);
+
+        const encoded = await wallet.core.crypto.encode(testSub.topic, {
+          ...message,
+          // Set different JSONRPC ID
```
Comment (bkrem) [inline packages/notify-client/test/clients.spec.ts:887]: Was trying to work out why this is being done at first, so adding the (presumed) reasoning:

```suggestion
          // Set different JSONRPC ID to avoid the relayer deduping the message based on JSON-RPC ID
```

**Review by bkrem — APPROVED** [review]

**Comment by bkrem** [comment]
Nice work here 🙌 

* One suggestion around perf optimisation
* Bit concerned about the arbitrary durations in the tests

File: packages/notify-client/test/clients.spec.ts (hunk 4 above)
Comment (devceline) [reply packages/notify-client/test/clients.spec.ts:887]: This is explained in the comments below, wouldn't adding this be redundant?

File: packages/notify-client/test/clients.spec.ts (hunk 3 above)
Comment (devceline) [reply packages/notify-client/test/clients.spec.ts:872]: I tried to think of something else to use but nothing happens between the ``RELAYER_EVENTS.message` event and the message being ignored. 

- `emit RELAYER_EVENTS.message`
- `onNotifyMessageRequest`
- some minor logs 
- message is ignored

The only thing I thought of was finding some way to capture messages from the logger but felt that was a bit hacky. As it stands, this test can't flake and fail randomly because it's checking for the lack of something. Worst case is a false positive, which is still bad undoubtedly.

File: packages/notify-client/src/controllers/engine.ts (hunk 1 above)
Comment (devceline) [reply packages/notify-client/src/controllers/engine.ts:625]: Message are indexed via payload ID. My initial solution was to change how messages are indexed but changing that is not backwards compatible.

Another solution I thought was just adding a secondary hashmap that just checks for existence of messages. Just a simple `string -> int` hashmap with message id and time received. 

However, I didn't know if that would be premature optimization so I just left it as is.

If you think it's worth it, I can go ahead with my second solution which is backwards compatible.

File: packages/notify-client/test/clients.spec.ts (hunk 3 above)
Comment (bkrem) [reply packages/notify-client/test/clients.spec.ts:872]: Ok I get it now thank you 👍

Proper way to detect this would be for the relayer to emit on ignored event, but that’s another story entirely. We can raise with Gancho if we keep seeing the need for it in places.

File: packages/notify-client/test/clients.spec.ts (hunk 4 above)
Comment (bkrem) [reply packages/notify-client/test/clients.spec.ts:887]: Action (override ID) and intent (avoid eager deduping) is clear but not mechanism (relayer will interfere here if not handled).

Not super important but felt this may be unclear to someone reading this if they’re not intimately familiar with the relayer logic)

Comment (devceline) [reply packages/notify-client/test/clients.spec.ts:887]: Ah I didn't consider that, Yep I'll commit your suggestion
